### PR TITLE
Better Makefile target parsing

### PR DESCRIPTION
**Pull request by adamjstewart**
While working on the `cryptopp` package, I ran into a `GNUmakefile` that didn't have a `^check:` or `^test:` target, it had a `^test check:` target. Apparently Makefiles can contain [multiple targets](https://www.gnu.org/software/make/manual/html_node/Multiple-Targets.html) that act as aliases. They can also have a space between the target and the colon. All of these things break our previous Makefile parsing logic.

I started playing around with regexes, but things quickly became complicated. I discovered an easier way to use `make` to query whether or not a specific target exists.

### Pros

More reliable Makefile parsing.

### Cons

This approach only works for GNU Make, it fails for BSD Make. I'm not sure how common BSD Make is (macOS and Linux both use GNU Make). The downside to this approach is that if someone runs `spack install --test` on a system that uses BSD Make, it will always think that a `check` target exists and run it even if it doesn't. The solution would either be to install GNU Make or to re-run `spack install` without `--test` if the test crashes.

P.S. I found a similar feature for Ninja and implemented that as well.

### CODE REVIEW

File: lib/spack/spack/package.py [hunk 1]
```diff
@@ -1245,43 +1245,68 @@ def do_fake_install(self):
...
-        except IndexError:
+        make = inspect.getmodule(self).make
+
+        # Check if we have a Makefile
+        for makefile in ['GNUmakefile', 'Makefile', 'makefile']:
+            if os.path.exists(makefile):
+                break
+        else:
             tty.msg('No Makefile found in the build directory')
             return
```
Comment (adamjstewart) [inline lib/spack/spack/package.py]: This block isn't actually needed anymore, it is only used in the final output message. Do you think it's still worth keeping?

**Comment by davydden** [comment]
> The solution would either be to install GNU Make or to re-run spack install without --test if the test crashes.

or you can fallback to the old code with regex if `make` is not a GNU make.

**Comment by adamjstewart** [comment]
Is the BSD Make executable actually called `make`? When I install the `bmake` package, I get an executable called `bmake`. I don't have access to a BSD system...

**Comment by citibeth** [comment]
BSD make was the original, and its executable was called `make`.  GNU Make was a clone, and its executable was also called `make`; or sometimes `gmake`.

I'm not against listing GNU make as a requirement; especially since it's already part of Linux and macOS, the two target platforms we run on.

**Comment by adamjstewart** [comment]
Ping @alalazo 

@tgamblin may also be necessary to make an executive decision as to how important non-GNU Make support is. If we really want to be careful, I can implement @davydden's suggestion and fall back on regexes.

File: lib/spack/spack/package.py [hunk 2]
```diff
@@ -1245,43 +1245,68 @@ def do_fake_install(self):
...
+        #
+        # 2     The exit status is two if make encounters any errors.
+        #       It will print messages describing the particular errors.
+        #
+        # 1     The exit status is one if you use the '-q' flag and make
+        #       determines that some target is not already up to date.
+        #
+        # https://www.gnu.org/software/make/manual/html_node/Running.html
+        #
+        # NOTE: This only works for GNU Make, not NetBSD Make.
```
Comment (scheibelp) [inline lib/spack/spack/package.py]: When you say:

> The downside to this approach is that if someone runs spack install --test on a system that uses BSD Make, it will always think that a check target exists and run it even if it doesn't.

How does this compare to the previous behavior for BSD make in Spack?

What is the exit status if you run this for BSD make? Can you distinguish a failing BSD make from a GNU make that is failing because the target doesn't exist?

Comment (adamjstewart) [reply lib/spack/spack/package.py]: Previously, we would grep the file for a `^check:` target. It didn't matter if it was BSD or GNU Make, it would behave identically. The problem was that this regex isn't powerful enough to capture all possible `check` targets (see first paragraph of PR description).

The current behavior is that it is guaranteed to work perfectly for GNU Make but doesn't work at all for BSD Make.

With BSD Make, there is no separate exit status for "this target doesn't exist", it's the same exit status as "the target has not yet been built", so there is no way to distinguish by exit status alone. We could fall back on a regex if we need to support BSD Make, although I've never encountered a system that uses BSD Make.

**Comment by adamjstewart** [comment]
Ping @alalazo

File: lib/spack/spack/package.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/package.py]: What about something like:
```python
if not any(os.path.exists(makefile) for makefile in ['GNUmakefile', 'Makefile', 'makefile']):
    tty.msg('No Makefile found in the build directory')
    return
```
?

Comment (adamjstewart) [reply lib/spack/spack/package.py]: I personally find this more readable, and you suggestion will need to be split onto multiple lines anyway. Plus, this results in `makefile` being set to the Makefile name, whereas I don't know if your suggestion does.

**Comment by adamjstewart** [comment]
There are alternatives to this PR, or things we could try in the case of BSD Make:
https://unix.stackexchange.com/questions/230047/how-to-list-all-targets-in-make
https://gist.github.com/pvdb/777954

I would have to see how many of those flags work across all versions of Make though.

File: lib/spack/spack/package.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/package.py]: Fair enough. Readability is subjective.

**Comment by adamjstewart** [comment]
Ping

**Comment by scheibelp** [comment]
> We could fall back on a regex if we need to support BSD Make

On account of that I'm inclined to merge this. If I understand correctly this would mean that for GNU makefiles without a check target you'd end up doing the `make check` and the regex (that's not a problem IMO just making sure I understand how that update would work).

I'll get a chance to read through this again by the end of Tuesday 7/17.

**Comment by adamjstewart** [comment]
@scheibelp Here's a summary of the previous and current behavior:

### Before

Identical behavior for GNU Make and BSD Make.

No false positives, many false negatives.

Ignored presence of `GNUmakefile`. Ignored situations like the following:

```make
check test:
test check:
check :
foo check bar:
```

### After

Different behavior for GNU Make and BSD Make.

No false positives or false negatives for GNU Make, 100% false positive rate for BSD Make.

I'm not sure how common non-GNU Make is. If you are concerned, we _could_ add logic to fall back on regexes if non-GNU Make is detected. If anyone can think of a better regex than `^check:`, let me know.

**Comment by scheibelp** [comment]
>> If I understand correctly this would mean that for GNU makefiles without a check target you'd end up doing the make check and the regex (that's not a problem IMO just making sure I understand how that update would work).

> 100% false positive rate for BSD Make.

Got it, I was incorrect in https://github.com/spack/spack/pull/8223#issuecomment-404991354: when you say "fall back to regex", we'd need to know that we are dealing with BSD make in the first place. 

`make -v` reports "GNU Make" for me, which is satisfactory if it is dependable across GNU make versions, and if BSD make doesn't report something similar (it looks like `-v` doesn't retrieve version info at all for BSD make: https://www.freebsd.org/cgi/man.cgi?make). I think that works well enough that we could at least add that information to an error message if `make(target)` in `_if_make_target_execute` fails.

In other words I wanted to make sure we can figure out whether we are using BSD or GNU make if we need to (later, not in this PR), and based on the above I think we can. If you agree then I'm fine merging this as is (I'll get a chance to do that tomorrow).

**Comment by adamjstewart** [comment]
Yes, checking the output of `make -v` should be sufficient to determine whether we are using GNU Make if we need to.
